File: aznavrail/src/test/java/com/hereliesaz/aznavrail/tutorial/AzCalloutPlacementTest.kt
package com.hereliesaz.aznavrail.tutorial

import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/** Verifies the non-blocking callout placement: near the target, off obstacles, inside the safe area. */
class AzCalloutPlacementTest {

    private val safe = Rect(0f, 0f, 1000f, 2000f)
    private val size = Size(240f, 100f)

    @Test
    fun `overlapArea computes the intersection`() {
        assertEquals(25f, overlapArea(Rect(0f, 0f, 10f, 10f), Rect(5f, 5f, 15f, 15f)))
        assertEquals(0f, overlapArea(Rect(0f, 0f, 10f, 10f), Rect(20f, 20f, 30f, 30f)))
    }

    @Test
    fun `prefers below the target when there is room`() {
        val target = Rect(400f, 400f, 600f, 500f)
        val r = placeCallout(target, size, obstacles = emptyList(), safe = safe)
        assertEquals(400f, r.left)
        assertEquals(508f, r.top) // target.bottom + gap(8)
    }

    @Test
    fun `never covers the target`() {
        val target = Rect(400f, 400f, 600f, 500f)
        val r = placeCallout(target, size, obstacles = emptyList(), safe = safe)
        assertEquals(0f, overlapArea(r, target))
    }

    @Test
    fun `flips above when below would leave the safe area`() {
        val target = Rect(400f, 1850f, 600f, 1950f)
        val r = placeCallout(target, size, obstacles = emptyList(), safe = safe)
        assertEquals(1742f, r.top) // target.top - height - gap
        assertTrue(r.bottom <= safe.bottom)
        assertEquals(0f, overlapArea(r, target))
    }

    @Test
    fun `avoids an obstacle blocking the preferred side`() {
        val target = Rect(400f, 400f, 600f, 500f)
        val obstacleBelow = Rect(380f, 500f, 680f, 700f)
        val r = placeCallout(target, size, obstacles = listOf(obstacleBelow), safe = safe)
        assertEquals(0f, overlapArea(r, obstacleBelow))
        assertEquals(0f, overlapArea(r, target))
    }

    @Test
    fun `stays fully inside the safe area`() {
        val target = Rect(950f, 50f, 999f, 99f) // hard against the right edge
        val r = placeCallout(target, size, obstacles = emptyList(), safe = safe)
        assertTrue(r.left >= safe.left && r.right <= safe.right)
        assertTrue(r.top >= safe.top && r.bottom <= safe.bottom)
    }

    @Test
    fun `untargeted callouts spread off each other`() {
        val placed = Rect(0f, 0f, 240f, 100f) // a callout already placed top-left
        val r = placeCallout(target = null, size = size, obstacles = listOf(placed), safe = safe)
        assertEquals(0f, overlapArea(r, placed))
        assertTrue(r.left >= safe.left && r.right <= safe.right && r.top >= safe.top && r.bottom <= safe.bottom)
    }
}

File: aznavrail-react/src/__tests__/AzCalloutPlacement.test.ts
import { placeCallout, overlapArea } from '../guidance/AzCalloutPlacement';
import type { AzShapeBounds } from '../guidance/AzStatus';

/** Parity with the Kotlin AzCalloutPlacementTest. */

const safe: AzShapeBounds = { left: 0, top: 0, width: 1000, height: 2000 };
const size = { width: 240, height: 100 };

describe('placeCallout', () => {
  it('overlapArea computes the intersection', () => {
    expect(overlapArea({ left: 0, top: 0, width: 10, height: 10 }, { left: 5, top: 5, width: 10, height: 10 })).toBe(25);
    expect(overlapArea({ left: 0, top: 0, width: 10, height: 10 }, { left: 20, top: 20, width: 10, height: 10 })).toBe(0);
  });

  it('prefers below the target when there is room', () => {
    const r = placeCallout({ left: 400, top: 400, width: 200, height: 100 }, size, [], safe);
    expect(r.left).toBe(400);
    expect(r.top).toBe(508);
  });

  it('never covers the target', () => {
    const t = { left: 400, top: 400, width: 200, height: 100 };
    expect(overlapArea(placeCallout(t, size, [], safe), t)).toBe(0);
  });

  it('flips above when below would leave the safe area', () => {
    const r = placeCallout({ left: 400, top: 1850, width: 200, height: 100 }, size, [], safe);
    expect(r.top).toBe(1742);
    expect(r.top + r.height).toBeLessThanOrEqual(safe.height);
  });

  it('avoids an obstacle blocking the preferred side', () => {
    const t = { left: 400, top: 400, width: 200, height: 100 };
    const ob = { left: 380, top: 500, width: 300, height: 200 };
    const r = placeCallout(t, size, [ob], safe);
    expect(overlapArea(r, ob)).toBe(0);
    expect(overlapArea(r, t)).toBe(0);
  });

  it('stays fully inside the safe area', () => {
    const r = placeCallout({ left: 950, top: 50, width: 49, height: 49 }, size, [], safe);
    expect(r.left >= 0 && r.left + r.width <= 1000).toBe(true);
    expect(r.top >= 0 && r.top + r.height <= 2000).toBe(true);
  });

  it('untargeted callouts spread off each other', () => {
    const placed = { left: 0, top: 0, width: 240, height: 100 };
    const r = placeCallout(null, size, [placed], safe);
    expect(overlapArea(r, placed)).toBe(0);
  });
});
